server/user_activity.go

```diff
@@ -0,0 +1,154 @@
+package main
+
+import (
+	"time"
+
+	"github.com/mattermost/mattermost-server/model"
+)
+
+type userChannelActivity = map[string]map[string]int64 //map[userID]map[channelID]activity
+
+// getActivity returns user activity in channels for every user from the beginning of times.
+// For now user activity is the number of posts posted in a channel.
+func (p *Plugin) getActivity() (userChannelActivity, error) {
+	previousTimestamp, err := p.retreiveTimestamp()
+	if err != nil {
+		return nil, err
+	}
+	timestampNow := time.Now().Unix()
+	activitySince, appErr := p.getActivitySince(previousTimestamp) // TODO what about the posts that where added between those lines?
+	if appErr != nil {
+		return nil, appErr
+	}
+	activityUntil, err := p.retreiveUserChannelActivity()
+	if err != nil {
+		return nil, err
+	}
+
+	activityUnion(activitySince, activityUntil)
+	if err = p.saveTimestamp(timestampNow); err != nil {
+		return nil, err
+	}
+	return activitySince, nil
+}
+
+// getActivitySince returns activities in channels for every user.
+// if since equals to -1 method returns activity score based on
+// all posts of the user, otherwise it score is based on posts since certain time.
+func (p *Plugin) getActivitySince(since int64) (userChannelActivity, *model.AppError) {
+	activity := make(userChannelActivity)
+	channels, err := p.GetAllChannels()
+	if err != nil {
+		return nil, err
+	}
+	for channelID := range channels {
+		var activityForChannel userChannelActivity
+		if since == -1 {
+			activityForChannel, err = p.getActivityForChannel(channelID)
+		} else {
+			activityForChannel, err = p.getActivityForChannelSince(channelID, since)
+		}
+		if err != nil {
+			return nil, err
+		}
+		activityUnion(activity, activityForChannel)
+	}
+	return activity, nil
+}
+
+func (p *Plugin) getActivityForChannel(channelID string) (userChannelActivity, *model.AppError) {
+	activity := make(userChannelActivity)
+	page := 0
+	perPage := 100
+	for {
+		posts, err := p.API.GetPostsForChannel(channelID, page, perPage)
+		if err != nil {
+			return nil, err
+		}
+		if len(posts.Order) == 0 {
+			break
+		}
+		pageActivity := getActivityFromPosts(posts, channelID)
+		activityUnion(activity, pageActivity)
+		page++
+	}
+	return activity, nil
+}
+
+func (p *Plugin) getActivityForChannelSince(channelID string, since int64) (userChannelActivity, *model.AppError) {
+	posts, err := p.API.GetPostsSince(channelID, since)
```

Comment:
`GetPostsSince` also returns posts that were updated in that time period, so you might want to double check that their `CreateAt` is greater than `since` in `getActivityFromPosts`

Reply:
I missed that, but as I think of it can't we assume that update is also an activity? Not as important as a post but still. Probably we might want to add more activities: likes, clicks on links...

Reply:
Yeah, that makes sense that we might want to track more activity other than just posting.


There's one potential problem with this though. Posts aren't always updated by the user that made the post, so there could be cases where one user's activity "modifies" another user's message which is then treated as activity for the other user. A post's UpdateAt timestamp gets changed when someone comments on the post or reacts to the post, so if UserA makes a post and UserB either comments on it or reacts to it, this function will see it as activity from UserA.

This probably isn't too much of a deal, but perhaps we want to filter `posts` for ones that have a CreateAt or EditAt greater than the since timestamp or at least add a comment to come back and fix this later